google/auth/credentials.py

```diff
@@ -157,11 +174,22 @@ def before_request(self, request, method, url, headers):
         # pylint: disable=unused-argument
         # (Subclasses may use these arguments to ascertain information about
         # the http request.)
-        if not self.valid:
+
+        if self.token_state == TokenState.STALE:
+            if self._use_non_blocking_refresh:
+                self._refresh_worker.start_refresh(self, request)
+            else:
+                self.refresh(request)
+
+        if self.token_state == TokenState.INVALID:
             self.refresh(request)
```

Comment:
Multiple requests may end up doing this refresh right?
Can we have a lock on this sync refresh?

Reply:
The code is using a bounded queue which should reduce duplicate work. There is a risk that there is a small amount of duplicate work, if two threads queue work at the same time. The queue itself is thread safe + has locks.

I believe using the queue to reduce duplicate work is acceptable, and that multiple threads will perform the refresh occasionally. 

There should be no data corruption due to this